managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -2451,15 +2451,15 @@ private void trimConsumedLedgersInBackground() {
 
     @Override
     public void trimConsumedLedgersInBackground(CompletableFuture<?> promise) {
-        executor.execute(() -> internalTrimConsumedLedgers(promise));
+        scheduledExecutor.execute(() -> internalTrimConsumedLedgers(promise));
     }
 
     public void trimConsumedLedgersInBackground(boolean isTruncate, CompletableFuture<?> promise) {
```

Comment:
Both tasks `trim ledgers` and `internalAsyncAddEntry` will try to acquire the lock `synchronized(ml)`, If the task `trim ledgers` running in the thread `executor`, it avoids lock contention.

Then if we make the task `trim ledgers` run in another thread, the publish task will also wait for the lock `synchronized(ml)`. Maybe we need improve this scenario first

Reply:
Maybe we can resolve this busy thread problem first, because the synchronized symbol only affects one ManagedLedger object, but one bk main executor thread may handle a lot of trim tasks for different ManagedLedger objects.

Reply:
Why not in two steps?
Thread pool isolation is necessary, trim should not run in the main thread pool
Next, let's optimize the time-consuming of trim

Reply:
@315157973 

Now the tasks `publish` and `trim ledgers` all run in the thread `executor`( it is a single thread pool, and the tasks of same topic will be executed in the same thread<sup>[L-1]</sup> ), right?<sup>[Q-1]</sup>

---

if yes<sup>[Q-1]</sup>, the task `trim ledgers` blocks the task `publish`, which this PR tries to fix, right?<sup>[Q-2]</sup>

---
if yes<sup>[Q-2]</sup>, But both tasks `trim ledgers`<sup>[L-2]</sup> and `internalAsyncAddEntry`<sup>[L-3]</sup> will try to acquire the lock `synchronized(ml)`, so the task `trim ledgers` is still blocking the task `publish` even if the patch is merged, right?[Q-3]</sup>

---
**[L-1]**: https://github.com/apache/pulsar/blob/master/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L348
**[L-2]**:  https://github.com/apache/pulsar/blob/master/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L2608
**[L-3]**: https://github.com/apache/pulsar/blob/master/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L798

Reply:
This lock is segmented and does not block each producer for a long time. 
But if they are in the same thread pool, the thread pool will be occupied for a long time

Reply:
> This lock is segmented and does not block each producer for a long time.

Make sense

Reply:
> if yes[Q-1], this patch doesn't have much effect on this issue, right?[Q-2]

 Assuming that one trim task increase 0.01ms,  if there are tens of thousands of partitions, it will increase a lot

Reply:
@315157973 

> Assuming that one trim task increase 0.01ms, if there are tens of thousands of partitions, it will increase a lot.
Especially when there are many small packets sent to the cluster， this situation will get worse

I agree with you. But I'm wondering if the current changes will solve the issue.
![截屏2023-07-03 22 29 13](https://github.com/apache/pulsar/assets/25195800/d681cb43-bc05-4a4e-ab46-1a0c6892eea2)

In the picture above, the logic of `trim ledgers` can be split into two sections: Verification(not in the lock block) and Execution(in the lock block), and there have some async tasks in Section Execution which will run in other threads. The three tasks execute as the flow below.

| Without lock | In lock | In other threads |
| --- | --- | --- |
| Verification | Memory changes | Persist changes |

Since there is the lock `synchronized(ml)`, we can only reduce the executor thread cost of the task `Verification`, which is very simple. It only tries to acquire the lock and is done if it fails.

**(Highlight)** However, this change also makes the scramble of the lock `synchronized(ml)` more frequent, which can worsen performance.

For example, the `topic-a` and `topic-b` were assigned the same thread: `executor-1.` And the task `trim ledgers` will be executed in the thread `scheduled executor`, these tasks will be executed as flow below

| time |`executor-1` | `scheduled executor` |
| --- | --- | --- |
| 1 | publish of `topic-a` |
| 2 | publish of `topic-b` |
| 3 | publish of `topic-b` | start `trim ledgers` of `topic-a` |
| 4 | publish of `topic-b` | get the ML lock of `topic-a` |
| 5 | publish of `topic-b` |
| 6 | publish of `topic-a` |
| 7 | waiting the lock, all publish of `topic-a` and `topic-b` will be blocked util the task `trim ledgers` of `topic-a` is complete |

So in step-4, if the task `trim ledgers` of `topic-a` is executed, the performance will be better; if the task `trim ledgers` of `topic-a` is executed, the performance will be worse. If more and more topics are assigned to the same thread, the performance will be better(maybe we can add a config to disable this feature, it is helpful for the users who have few topics).

Reply:
Users with fewer partitions submit fewer tasks to the thread pool

Reply:
This change can improve the performance if there are a lot of topics in the broker and there are few threads in the BK client thread pool.

Make sense to me.